### PR TITLE
Rework the package -- Take 1

### LINKED ISSUE #81: Minor issues when wrapping libav
- [x] Strip `UL`/`LL` from long constants (and maybe others)
- [ ] Recognize `INT64_C`/`UINT64_C` (and related?) macros
- [x] Type enum values as `Cint` or `Uint32`, to match enum type

### LINKED ISSUE #126: Wrap support for getOffsetOf
http://clang.llvm.org/doxygen/group__CINDEX__TYPES.html#gab543536d5c18efb3e23a1b7903fb494d

per https://groups.google.com/d/msg/julia-dev/0Tui0v3lgts/XvkZYA3APuoJ

### LINKED ISSUE #136: missing argument names for Typedef'd args
Root cause is that the original names seem to be missing from the `ParmDecl`s. Not sure what the deal is yet, but could be a libclang issue. Output still works fine with the generic "arg1", "arg2", but it would be nice to give the intended name.

### LINKED ISSUE #195: Function and other signatures in comments
For spot-checking and learning it would be nice to have an option to include original C signatures as comments (or docstrings?) next to the ccall and possibly other declarations.

### LINKED ISSUE #224: no method `value(::CLPackedAttr)`
```
ERROR: LoadError: MethodError: no method matching value(::CLPackedAttr)
Closest candidates are:
  value(!Matched::CLEnumConstantDecl) at /home/vchuravy/.julia/packages/Clang/NONKM/src/cursor.jl:246
Stacktrace:
 [1] wrap!(::DefaultContext, ::CLEnumDecl) at /home/vchuravy/.julia/packages/Clang/NONKM/src/wrap_c.jl:97
 [2] run(::WrapContext) at /home/vchuravy/.julia/packages/Clang/NONKM/src/compat.jl:122
 [3] top-level scope at none:0
````

It seems to be just:

```
typedef enum {
    /* Operation completed successfully */
    UCS_OK                         =   0,

    /* Operation is queued and still in progress */
    UCS_INPROGRESS                 =   1,

    /* Failure codes */
    UCS_ERR_NO_MESSAGE             =  -1,
    UCS_ERR_NO_RESOURCE            =  -2,
    UCS_ERR_IO_ERROR               =  -3,
    UCS_ERR_NO_MEMORY              =  -4,
    UCS_ERR_INVALID_PARAM          =  -5,
    UCS_ERR_UNREACHABLE            =  -6,
    UCS_ERR_INVALID_ADDR           =  -7,
    UCS_ERR_NOT_IMPLEMENTED        =  -8,
    UCS_ERR_MESSAGE_TRUNCATED      =  -9,
    UCS_ERR_NO_PROGRESS            = -10,
    UCS_ERR_BUFFER_TOO_SMALL       = -11,
    UCS_ERR_NO_ELEM                = -12,
    UCS_ERR_SOME_CONNECTS_FAILED   = -13,
    UCS_ERR_NO_DEVICE              = -14,
    UCS_ERR_BUSY                   = -15,
    UCS_ERR_CANCELED               = -16,
    UCS_ERR_SHMEM_SEGMENT          = -17,
    UCS_ERR_ALREADY_EXISTS         = -18,
    UCS_ERR_OUT_OF_RANGE           = -19,
    UCS_ERR_TIMED_OUT              = -20,
    UCS_ERR_EXCEEDS_LIMIT          = -21,
    UCS_ERR_UNSUPPORTED            = -22,
    UCS_ERR_REJECTED               = -23,

    UCS_ERR_FIRST_LINK_FAILURE     = -40,
    UCS_ERR_LAST_LINK_FAILURE      = -59,
    UCS_ERR_FIRST_ENDPOINT_FAILURE = -60,
    UCS_ERR_LAST_ENDPOINT_FAILURE  = -79,
    UCS_ERR_ENDPOINT_TIMEOUT       = -80,

    UCS_ERR_LAST                   = -100
} __attribute__((packed)) ucs_status_t ;

```

### LINKED ISSUE #233: Unnamed struct members are skipped
Unnamed nested structures/unions are currently skipped.

Given
```C
struct Meow {
    int blah;
    union {
        float either1;
        int either2;
    };
    float moo;
};
```
the lib generates the following Julia code:
```Julia
struct Meow
    blah::Cint
    moo::Cfloat
end
```

The problem goes away when giving the member a name. Maybe generating a member called something along the lines of "anon_0" for these unnamed nested items would be best?

### LINKED ISSUE #238: Potential wrong alignments and sizes of structs with C unions
There is an issue in the way Clang.jl wraps C unions by just picking the largest alternative.
The problem occurs when a union is inside a struct and when such union contains an alternative which, while not being the largest one, is composed of elements with greater word sizes than all the other alternatives or other elements of the struct.

See this example:

```c
struct struct_with_union
{
  float x;
  union {
    double y;
    float z[10];
  } foo;
} struct_with_union;
```
Clang will wrap that in Julia by something like (names not completely right):
```julia
struct ANONYMOUS___
  z::NTuple{10, Float32}
end
struct struct_with_union
  x::Float32
  foo::ANONYMOUS___
end
```

The size of `struct_with_union` in Julia will be 44 bytes: 4 for `x` plus 10*4 for `z` (the largest in the union). 

However, the actual size of the struct in C is 48 bytes: 10*4 for `z` plus 4 for `x`, plus padding of 4 between `x` and `z`.
That padding comes, AFAIK, to the fact that the largest word possible in the struct is the 8-byte double from the `y` alternative and thus all the elements get aligned to the 8-byte boundaries.

### LINKED ISSUE #240: Some header files entirely skipped
My `libjpeg` wrapper completes successfully, but on loading the output hits errors because of missing consts, the first of which is defined in `jmorecfg.h` and I noticed the following warnings. Why are those provided headers being skipped?
```
┌ Warning: not wrapping CLCursor (CLLastPreprocessing) jconfig.h
└ @ Clang ~/.julia/packages/Clang/CiPzM/src/wrap_c.jl:443
┌ Warning: not wrapping CLCursor (CLLastPreprocessing) jmorecfg.h
└ @ Clang ~/.julia/packages/Clang/CiPzM/src/wrap_c.jl:443
```

Wrapper code: https://github.com/ianshmean/ImageIO.jl/blob/master/gen/libjpeg/wrap_libjpeg.jl

### LINKED ISSUE #279: Argument and type name equivalent
```
void flux_stat_watcher_get_rstat (flux_watcher_t *w,
                                  struct stat *stat, struct stat *prev)
```

Produces:

```
function flux_stat_watcher_get_rstat(w, stat, prev)
    ccall((:flux_stat_watcher_get_rstat, libflux_core), Cvoid, (Ptr{flux_watcher_t}, Ptr{stat}, Ptr{stat}), w, stat, prev)
end
```

Which fails with, because the type name `stat` is shadowed by the local variable name `stat`

x-ref: https://github.com/flux-framework/FluxRM.jl/pull/6

### LINKED ISSUE #280: CXType_FunctionProto
```
typedef void (flux_moderr_f)(const char *errmsg, void *arg);
```
produces:

```
120:# Skipping Typedef: CXType_FunctionProto flux_moderr_f
```

x-ref: https://github.com/flux-framework/FluxRM.jl/pull/6

### PR DESCRIPTION
**Pull request by Gnimuc**
Well, I ended up reworking the whole package. 

Features added by this PR:
- Add a `Generators` module for wrapper generation. The context now maintains a DAG of the original AST of the C code. This DAG can be used for CTU analysis. The old `wrap` methods are split into different passes, this makes it easier to add features to the generator, e.g. printing useful C comments(fix #195).
- Expose a generic way for users who would like to add translation rules for custom definitions. (fix #81)
- Forward declarations in C code can now be correctly handled by the generator(fix #240), mutual-referenced structs are de-referenced at the pointer field types.
- All tag-types that have a special layout(`__attribute__`/`bitfield`/`nested anonymous`) are generated as a memory-pool-like struct. The size is queried from libclang, so it guarantees that issues like #238 won't happen(*assume the code is system-independent*). (fix #238, fix #224)
- High-level API is simplified. With this PR, the new API function `create_context` only accepts 3 arguments(headers, compiler arguments, options) instead of 8 arguments. Additional options are loaded from a .toml file. By default, the generator prints a complete module to a single file.
- The package is now self-contained. 

In short, the goal is to make the wrapper generator easier to extend for new features. 

As mentioned above, this PR is breaking. But I guess it's not that hard to upgrade the old scripts. I'm going to manually submit upgrading PRs to the downstream repos, which can also serve as a validation of the correctness of this PR.

TODO list:
- [x] Add back basic macro support
- [x] Generate nested anonymous structs(fix #233)
- [x] Generate `getproperty`/`setproperty` for structs with special layout
- [ ] ~~Support vararg functions with the new `@ccall` macro(#17, #146)~~
- [ ] ~~Register `LibClang.jl` as a standalone package~~

Other bugfixes: fix #280, fix #279,  fix #136, fix #126

### CODE REVIEW

**Comment by ihnorton** [comment]
I'm trying this out, and it looks amazing ❤️

The only thing I've run in to so far is handling of `{u}int{32,64}_t` arguments -- they seem to be translated as bare `uint32_t` identifier, which doesn't exist (adding `const uint32_t = UInt32` etc. fixed it). I'm happy to open an issue if you prefer.

**Comment by Gnimuc** [comment]
hi @ihnorton, could you share the C code that triggers the issue? I simplified the rule for translating function return types, but this change may be wrong for some corner cases.

**Comment by ihnorton** [comment]
The signatures are [here](https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/sm/c_api/tiledb.h#L427-L443).

Looking closer, I see what is going on (overlooked the warning comment 🙃 ): these are all classified as `CXType_FunctionNoProto`. In the current `master` I believe this situation is handled here: https://github.com/JuliaInterop/Clang.jl/blob/01bf3e81aec6bec287a0794f3997666d8ea96db5/src/type.jl#L208-L222 (I had tried the same header a while back with an earlier `master`, and the return type was translated correctly as `UInt64` etc.)

**Comment by Gnimuc** [comment]
Thanks for the hint. Should now be fixed by https://github.com/JuliaInterop/Clang.jl/pull/278/commits/df520218b9b9583d04f40993d4f1c2c8e73ad721.

**Comment by ihnorton** [comment]
Works perfectly now, thanks!

**Comment by sairus7** [comment]
Will this rework fix this problem?
[Clang.jl often inserts a Cint when it isn't sure (in this case because they're types from another header file).](https://github.com/wsphillips/ImPlot.jl/issues/11#issuecomment-802977031)

**Comment by Gnimuc** [comment]
yep

2021年3月20日(土) 2:13 Viktor G. Gumennyy ***@***.***>:

**Comment by vrzh** [comment]
This rework effort is looking great! 

Is it on purpose that in `test/test.toml`, there is no "prologue" stuff included? otherwise `Ctime_t` will not be defined in the output: 

```
prologue_file_path = "../gen/prologue.jl"
```

Also, I have been unlucky that the C library I am trying to generate bindings for has a function which has a parameter named `Ptr`, which conflicts with the Julia `Ptr`. Generating the bindings succeeds, but when trying them I was getting:

```
ERROR: LoadError: could not evaluate ccall argument type (it might depend on a local variable)
```

I got around it by:

```julia
import Clang.Generators.RESERVED_WORDS
# ...
push!(RESERVED_WORDS, "Ptr")
```

Does this warrant enough of a use-case to expose a function/macro to add other words which may conflict? Similar to `@add_def` which helps to add extra definitions?

**Comment by Gnimuc** [comment]
> Is it on purpose that in `test/test.toml`, there is no "prologue" stuff included? otherwise `Ctime_t` will not be defined in the output:

I'm mainly running https://github.com/Gnimuc/GeneratorScripts for meaningful testing of the generated files, so only the generator's public API is tested for now. 

> Does this warrant enough of a use-case to expose a function/macro to add other words which may conflict? Similar to @add_def which helps to add extra definitions?

@vrzh, thanks for the feedback! Without adding new features, a better fix would be writing a rewriter for that specific function like [this one](https://github.com/JuliaInterop/Clang.jl/blob/f476f99659ecc162d9c46616ebe433531d5e5560/gen/generator.jl#L27-L65).

If you'd like to have this feature in Clang.jl, I'd suggest that we just add a new entry in the toml file like:
```toml
[codegen]
function_argument_conflict_symbols = ["Ptr"]  # should be empty by default
```
and prefix the argument names with "_" according to this vector in the codegen.jl.

**Comment by Gnimuc** [comment]
@vrzh, OK, I just implemented that in the latest commit, could you give it a try?

**Comment by vrzh** [comment]
> @vrzh, OK, I just implemented that in the latest commit, could you give it a try?

Doesn't seem to work for me. I am new to Julia so most likely my fault. I tried the following in a clean Julia environment.

`test.jl`:
```julia
using Pkg
Pkg.add(PackageSpec(name="Clang", rev="4fb6690"))

using Clang.Generators
using Clang.LibClang.Clang_jll

headers = [ "test.h" ]
options = load_options(joinpath(@__DIR__, "test.toml"))
args = [ "-I." ]

# import Clang.Generators.RESERVED_WORDS
# push!(RESERVED_WORDS, "Ptr")

ctx = create_context(headers, args, options)

build!(ctx)
```

`test.toml`:
```toml
[general]
library_name = "libtest"

[codegen]
function_argument_conflict_symbols = ["Ptr"] 
```

`test.h`:
```c
void f(int * x, int Ptr);
```

Output:
```julia
function f(x, Ptr)
    ccall((:f, libtest), Cvoid, (Ptr{Cint}, Cint), x, Ptr)
end
```

I agree with you that a rewriter is the best way to go (as you could also have C function names, C struct names, etc. which also clash), but an entry in the toml file as you suggest is the easiest for new users.

**Comment by Gnimuc** [comment]
@vrzh Oops! Should be fixed now.

> I agree with you that a rewriter is the best way to go (as you could also have C function names, C struct names, etc. which also clash), but an entry in the toml file as you suggest is the easiest for new users.

As C doesn't have any name mangling stuff, a common C library usually has its own manually mangled prefixes, so generally speaking, this rarely happens in practice.

**Comment by vrzh** [comment]
> @vrzh Oops! Should be fixed now.

Perfect! Works great, thanks a lot.

**Comment by Gnimuc** [comment]
This PR is too long, so let's merge this to master.
